### PR TITLE
Add _reshard HTTP API reference documentation

### PR DESCRIPTION
**Pull request by nickva**
Related to the main PR: https://github.com/apache/couchdb/pull/1972

### CODE REVIEW

File: src/api/server/common.rst [hunk 1]
```diff
@@ -1675,3 +1675,479 @@ You can verify the change by obtaining a list of UUIDs:
...
+        Content-Type: application/json
+
+        {
+            "ok": true
+        }
+
+.. http:get:: /_reshard/jobs
+    :synopsis: Retrieve information about all the resharding jobs on the cluster
+
+    Get information about the resharding job identified by ``jobid``.
```
Comment (flimzy) [inline src/api/server/common.rst]: I think this paragraph is misplaced, as `{jobid}` isn't mentioned here, but in the next section.

File: src/api/server/common.rst [hunk 2]
```diff
@@ -1675,3 +1675,479 @@ You can verify the change by obtaining a list of UUIDs:
...
+    :<json string type: Type of job. Currently only ``"split"`` is accepted.
+
+    :<json string db: Database to split. This is mutually exclusive with the
+                      ``"shard``" field.
+
+    :<json string node: Split shards on a particular node. This is an optional
+                        parameter. The value should be one of the nodes
+                        returned from the ``_membership`` endpoint.
+
+    :<json string range: Ssplit shards copies in the given range. The range
```
Comment (flimzy) [inline src/api/server/common.rst]:
```suggestion
    :<json string range: Split shard copies in the given range. The range
```

File: src/cluster/sharding.rst [hunk 3]
```diff
@@ -247,6 +247,10 @@ documents is mapped.
 Moving a shard
 --------------
 
+When moving shards or performing other shard manipulations on the cluster, it
+is adviseable to stop all resharding jobs on the cluster. See
```
Comment (flimzy) [inline src/cluster/sharding.rst]:
```suggestion
is advisable to stop all resharding jobs on the cluster. See
```

File: src/cluster/sharding.rst [hunk 4]
```diff
@@ -666,25 +670,207 @@ cluster do not host any replicas for newly created databases, by giving
...
+temporarily. Then, the state would be ``stopped`` and hopefully, there would be
+a reason or a comment in the value of the ``state_reason`` field. See
+:ref:`cluster/sharding/stop_resharding` for more details.
+
+The ``total`` number of jobs is important to keep an eye on because there is a
+maximum number of resharding jobs per node, and creating new jobs after the
+limit has been reached will result in an error. Before staring new jobs it's a
+good idea to remove already completed jobs. See :ref:`reshard configuration
+section <config/reshard>` for the default value of ``max_jobs`` parameter and
+how to adjusted if needed.
```
Comment (flimzy) [inline src/cluster/sharding.rst]:
```suggestion
how to adjust it, if needed.
```

File: src/cluster/sharding.rst [hunk 5]
```diff
@@ -666,25 +670,207 @@ cluster do not host any replicas for newly created databases, by giving
...
+      },
+      {
+        "ok": true,
+        "id": "001-fc72090c006d9b059d4acd99e3be9bb73e986d60ca3edede3cb74cc01ccd1456",
+        "node": "node3@127.0.0.1",
+        "shard": "shards/00000000-7fffffff/db1.1554242778"
+      }
+    ]
+
+The request returned three jobs. One job for each of the three copies.
```
Comment (flimzy) [inline src/cluster/sharding.rst]:
```suggestion
The request returned three jobs, one job for each of the three copies.
```

File: src/cluster/sharding.rst [hunk 6]
```diff
@@ -666,25 +670,207 @@ cluster do not host any replicas for newly created databases, by giving
...
+.. _cluster/sharding/stop_resharding:
+
+Stopping Resharding Jobs
+------------------------
+
+Resharding at the cluster level could be stopped and then restarted. This can
+be helpful to allow external tools which manipulate the shard map to avoid
+interfering with resharding jobs. To stop all resharding jobs on a cluster
+issue a ``PUT`` to ``/_reshard/state`` endpoint with the ``"state": "stopped"``
+key and value. Can also specify an optional note or reason for stopping.
```
Comment (flimzy) [inline src/cluster/sharding.rst]:
```suggestion
key and value. You can also specify an optional note or reason for stopping.
```

Would it be worth making this a stronger suggestion? "You should also specify..."?

File: src/api/server/common.rst [hunk 7]
```diff
@@ -1675,3 +1675,479 @@ You can verify the change by obtaining a list of UUIDs:
...
+    :>header Content-Type: - :mimetype:`application/json`
+
+    :<json string state: ``stopped`` or ``running``
+    :<json string state_reason: Optional string describing additional
+                                information or reason associated with the state
+
+    :>json boolean ok: ``true``
+
+    :code 200: Request completed successfully
+    :code 401: CouchDB Server Administrator privileges required
```
Comment (flimzy) [inline src/api/server/common.rst]: Is it worth mentioning 400 here?

File: src/api/server/common.rst [hunk 8]
```diff
@@ -1675,3 +1675,479 @@ You can verify the change by obtaining a list of UUIDs:
...
+
+    :<json string state: ``stopped`` or ``running``
+    :<json string state_reason: Optional string describing additional
+                                information or reason associated with the state
+
+    :>json boolean ok: ``true``
+
+    :code 200: Request completed successfully
+    :code 404: The job was not found
+    :code 401: CouchDB Server Administrator privileges required
```
Comment (flimzy) [inline src/api/server/common.rst]: Is it worth mentioning 400 here?

File: src/config/resharding.rst [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+        will start to fail with the ``max_jobs_exceeded`` error::
+
+             [reshard]
+             max_jobs = 25
+
+    .. config:option:: max_retries
+
+        How many times to retry shard splitting steps if they fail. For
+        example, if indexing or topping off fails, it will be retried up to
+        these many times before the whole resharding job fails::
```
Comment (flimzy) [inline src/config/resharding.rst]:
```suggestion
        this many times before the whole resharding job fails::
```

File: src/config/resharding.rst [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+
+             [reshard]
+             max_retries = 1
+
+    .. config:option:: retry_interval_sec
+
+        How to long to wait between subsequent retries::
```
Comment (flimzy) [inline src/config/resharding.rst]:
```suggestion
        How long to wait between subsequent retries::
```

File: src/config/resharding.rst [hunk 11]
```diff
@@ -0,0 +1,79 @@
...
+        utilized by the shard. When debugging or when extra safety is required,
+        this can be switched to ``false``::
+
+             [reshard]
+             delete_source = true
+
+    .. config:option:: update_shard_map_timeout_sec
+
+        How many seconds to wait for the shard map update operation to
+        complete. If there are a large number of shards db change waiting to
```
Comment (flimzy) [inline src/config/resharding.rst]:
```suggestion
        complete. If there is a large number of db shard changes waiting to
```

File: src/api/server/common.rst (hunk 7 above)
Comment (nickva) [reply src/api/server/common.rst]: It won't apply here because it's getting the state and resharding will have either a stopped or running state.

Comment (nickva) [reply src/api/server/common.rst]: Ah yes, for PUT it would make sense. Good call!

File: src/api/server/common.rst (hunk 8 above)
Comment (nickva) [reply src/api/server/common.rst]: Agree, good idea.

Comment (nickva) [reply src/api/server/common.rst]: Also noticed I needed to add one for POST to `_reshard/jobs`

**Comment by nickva** [comment]
@flimzy, I updated the PR. Thank you for the review! Good finds and great suggestions.

**Comment by flimzy** [comment]
LGTM!

**Comment by nickva** [comment]
Hmm, travis silliness. Build job finished green but never reported back to this commit so it still looks yellow here.

**Pull request merged**